[`pygrep_hooks`] Check blanket ignores via file-level pragmas (`PGH004`)

ajesipow opened the pull request:
## Summary

Should resolve https://github.com/astral-sh/ruff/issues/11454.

This is my first PR to `ruff`, so I may have missed something.

If I understood the suggestion in the issue correctly, rule `PGH004` should be set to `Preview` again.

## Test Plan

Created two fixtures derived from the issue.

JonathanPlasse commented:
The stable behavior should be unchanged, the added behavior should only be present in preview.
i.e. The ruff ecosystem results should not change for stable, only preview should have additional violations.
Otherwise, the users of stable will lose the PGH004 rule.

ajesipow commented:
Thanks for the swift feedback.

I added a check for `PreviewMode::Enabled` and added tests for the preview case.

Would you like me to also update the docs for `BlanketNOQA`?

charliermarsh commented on crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs:
```diff
@@ -85,7 +86,19 @@ pub(crate) fn blanket_noqa(
     diagnostics: &mut Vec<Diagnostic>,
     noqa_directives: &NoqaDirectives,
     locator: &Locator,
+    exemption: &Option<FileExemption>,
```
One downside here is that this is only going to flag the _first_ `# ruff: noqa` in the file. I assume if you add multiple such comments, only the first is flagged as a diagnostic?

charliermarsh replied:
We might need to do something like `NoqaDirectives`, whereby we parse all of these upfront, then use the parsed exemptions in `crates/ruff_linter/src/checkers/noqa.rs` (and here).

charliermarsh replied:
@ajesipow - Do you want to give that a try?

ajesipow replied:
Thanks for the ping @charliermarsh.
I haven't found the time yet after work this week, but today is a public holiday over here so I'll look into it today :)

ajesipow replied:
I just pushed the changes and adapted the tests again accordingly.

Looking forward to your feedback again.

ajesipow commented on crates/ruff_linter/src/noqa.rs:
```diff
@@ -275,26 +283,43 @@ pub(crate) fn rule_is_ignored(
     All,
     /// The file is exempt from the given rules.
-    Codes(Vec<NoqaCode>),
+    Codes(Vec<&'a NoqaCode>),
```
I left this as a `Vec`, but we could consider using a `HashSet` instead as we're checking if certain `NoqaCode`s are contained in a few places (one being the loop on line 54 in `checkers/noqa.rs`). 

It's probably fine though as I'd expect there will usually only be very few elements here and checking the small `Vec` is likely faster compared to the overhead of hashing every time (and building the `HashSet` in the first place).
We'd also have to derive `Hash` on `NoqaCode`.

charliermarsh commented:
Thanks @ajesipow, I will take a look.

charliermarsh reviewed (approved):
Thank you!
